src/videoTrackingHTML5.js

```diff
@@ -0,0 +1,248 @@
+
+    const processPauseMediaEvent = (e) => {
+        e.target.hasPaused = true;
+        var currentPlayTime = e.target.currentTime;
+        var videoTitle = getVideoName(e.target);
+        var isUserSeeking = e.target.seeking;
+        if (e.target.currentTime < e.target.duration && isUserSeeking == false) {
+            var eventData = {
+                eventType: "Pause",
+                videoTitle: videoTitle,
+                eventTimestamp: currentPlayTime,
+                currentVolume: e.target.volume
+            };
+            trackEvent(eventData);
+            e.target.hasPaused = true;
+        } else {
+           e.target.hasPaused = false;
+        }
+    };
+
+    const processSeekedMediaEvent = (e) => {
+        var isUserSeeking = e.target.seeking;
+        var eventData = {
+            videoTitle: getVideoName(e.target),
+            eventTimestamp: e.target.currentTime,
+            currentVolume: e.target.volume
+        };
+        if (e.target.currentTime < e.target.duration && isUserSeeking == false) {
+            if (e.target.hasPaused == true){
+                eventData.eventType = "Seeked when paused";
+                trackEvent(eventData);
+            } else if (e.target.hasPlayed == false && e.target.hasPaused == false) {
+                eventData.eventType = "Seeked before playback";
+                trackEvent(eventData);
+            } else if (e.target.hasPlayed == true && e.target.hasPaused == false && e.target.hasReplayed == false && e.target.hasEnded == true) {
+                eventData.eventType = "Seeked after finished watching";
+                trackEvent(eventData);
+                //workaround for the fact that after resume type of "Play", "seeking during playback" is automatically triggered
+            } else if (e.target.hasPlayed == true && e.target.hasPaused == false && e.target.hasReplayed == false) {
+                eventData.eventType = "Seeked during playback";
+                trackEvent(eventData);
+            } 
+             else if (e.target.hasReplayed == true){
+                e.target.hasReplayed = false;
+            }
+        }
+    };
+
+    const processEndedMediaEvent = (e) => {
+        e.target.hasPaused = false;
+        e.target.hasEnded = true;
+        var eventData = {
+            eventType: "Watched",
+            videoTitle: getVideoName(e.target),
+            eventTimestamp: e.target.currentTime,
+            currentVolume: e.target.volume
+        };
+        trackEvent(eventData);
+    };
+
+    const handleTabUnloadDuringVideoPlay = (e) => {
+        mediaElements = document.querySelectorAll("video");
+        mediaElements.forEach((mediaElement) => {
+            //4Patryk: I didn't have better idea for getting around eslint, it has issue with me checking for my own hasPlayed property on an object (it's added at the beginning of code execution)
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
```

Comment:
It's typescript linter by we cannot extend that type he has a problem with (because you are doing object mutation) in jS file. Leave it as it is for now.

Reply:
thanks for explanation, now as I think of it I actually could've omitted that object mutation (adding my own properties) but it would demand a different architecture of the solution - maybe in the future :)